pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageCrypto.java

```diff
@@ -0,0 +1,512 @@
+     */
+    public boolean removeKeyCipher(String keyName) {
+
+        if (keyName == null) {
+            return false;
+        }
+        encryptedDataKey.remove(keyName);
+        return true;
+    }
+
+    /*
+     * Encrypt the payload using the data key and update message metadata with the keyname & encrypted data key
+     * 
+     * @param encKeys One or more public keys to encrypt data key
+     * 
+     * @param msgMetadata Message Metadata
+     * 
+     * @param payload Message which needs to be encrypted
+     * 
+     * @return encryptedData if success
+     */
+    public ByteBuf encrypt(ConcurrentOpenHashSet<String> encKeys, MessageMetadata.Builder msgMetadata, ByteBuf payload)
+            throws PulsarClientException {
+
+        if (encKeys.isEmpty()) {
+            return payload.retain();
+        }
+
+        // Update message metadata with encrypted data key
+        encKeys.forEach(keyName -> {
+            msgMetadata.addEncryptionKeys(KeyByteValue.newBuilder().setKey(keyName)
+                    .setValue(ByteString.copyFrom(encryptedDataKey.get(keyName))).build());
+        });
+
+        // Create gcm param
+        // TODO: Replace random with counter and periodic refreshing based on timer/counter value
+        secureRandom.nextBytes(iv);
+        GCMParameterSpec gcmParam = new GCMParameterSpec(ivLen * 8, iv);
+
+        // Update message metadata with encryption param
+        msgMetadata.setEncryptionParam(ByteString.copyFrom(iv));
+
+        ByteBuf targetBuf = null;
+        try {
+            // Encrypt the data
+            cipher.init(Cipher.ENCRYPT_MODE, dataKey, gcmParam);
+
+            ByteBuffer sourceNioBuf = payload.nioBuffer(payload.readerIndex(), payload.readableBytes());
+
+            int maxLength = cipher.getOutputSize(payload.readableBytes());
+            targetBuf = PooledByteBufAllocator.DEFAULT.buffer(maxLength, maxLength);
+            ByteBuffer targetNioBuf = targetBuf.nioBuffer(0, maxLength);
+
+            int bytesStored = cipher.doFinal(sourceNioBuf, targetNioBuf);
+            targetBuf.writerIndex(bytesStored);
+
+        } catch (IllegalBlockSizeException | BadPaddingException | InvalidKeyException
+                | InvalidAlgorithmParameterException | ShortBufferException e) {
+
+            targetBuf.release();
+            log.error("Failed to encrypt message. {}", e);
+            throw new PulsarClientException.CryptoException(e.getMessage());
+
+        }
+        payload.release();
+        return targetBuf;
+    }
+
+    private boolean decryptDataKey(String keyName, byte[] encryptedDataKey, CryptoKeyReader keyReader) {
+
+        // Prefix "private-key." into key name and read the key value using callback
+        String privKeyName = "private-key." + keyName;
+        byte[] keyValue = keyReader.getKey(privKeyName);
+
+        // Convert key from byte to PivateKey
+        PrivateKey privateKey;
+        try {
+            privateKey = loadPrivateKey(keyValue);
```

Comment:
Should we cache this privateKey for a given key-name. Because, for a given key-name, value of the key will not be changed. Else we will end up calling `loadPrivateKey` for alternate message if we have multiple producers (because each producer has different dataKey which will fail at line: 479 and it forces to recalculate `dataKey`.